### PR TITLE
[RLlib] New API stack: (Multi)RLModule overhaul vol 02 (VPG RLModule, Algo, and Learner example classes).

### PR DESCRIPTION
**Pull request by sven1977**
New API stack: (Multi)RLModule overhaul vol 02 (VPG RLModule, Algo, and Learner example classes).
These new classes will replace the `core/testing/bc_...` classes and serve as:
- simple example implementations for things like: using a shared encoder in multi-agent.
- example for how to implement a custom Algorithm and AlgorithmConfig.
- example for yet another custom Learner.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/algorithms/classes/vpg.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        # Sync weights.
+        with self.metrics.log_time((TIMERS, SYNCH_WORKER_WEIGHTS_TIMER)):
+            self.env_runner_group.sync_weights(
+                from_worker_or_learner_group=self.learner_group,
+                policies=list(set(learner_results.keys()) - {ALL_MODULES}),
+                inference_only=True,
+            )
+
+        # Reduce and return collected metrics.
+        return self.metrics.reduce()
```
Comment (simonsays1980) [inline rllib/examples/algorithms/classes/vpg.py:107]: It's so simple now :)

File: rllib/examples/rl_modules/classes/vpg_rlm.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        self._policy_net = nn.Sequential(
+            nn.Linear(input_dim, hidden_dim),
+            nn.ReLU(),
+            nn.Linear(hidden_dim, output_dim),
+        )
+
+    @override(TorchRLModule)
+    def _forward_inference(self, batch, **kwargs):
+        with torch.no_grad():
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/vpg_rlm.py:29]: Nice!

File: rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            return self._common_forward(batch)
+
+    @override(RLModule)
+    def _forward_train(self, batch):
+        return self._common_forward(batch)
+
+    def _common_forward(self, batch):
+        # Features can be found in the batch under the "encoder_features" key.
+        features = batch["encoder_features"]
+        logits = self.policy_head(features)
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py:48]: `self._policy_head(features)`?

File: rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+        # Assert, we have the shared encoder submodule.
+        assert (
+            SHARED_ENCODER_ID in self._rl_modules
+            and isinstance(self._rl_modules[SHARED_ENCODER_ID], SharedTorchEncoder)
+            and len(self._rl_modules) > 1
+        )
+
+    @override(MultiRLModule)
+    def _run_forward_pass(self, forward_fn_name, batch, **kwargs):
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py:112]: This is still a strange form - passing in the `forward_fn_name`. Why is this needed? To have a single method to be overridden in inheritance?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Just a small nit and a deisgn question.

File: rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py (hunk 3 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py:48]: fixed

File: rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py (hunk 4 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py:112]: This will be gone in another PR that's already in (your) review :)

Comment (sven1977) [reply rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py:112]: Great question!
